dandiapi/api/tests/factories.py

```diff
@@ -116,32 +134,32 @@ def metadata(self):
         # Remove faked data that might conflict with the schema types
         for key in ['about']:
             metadata.pop(key, None)
+        if self.status == Version.Status.PUBLISHED:
+            now = datetime.datetime.now(datetime.UTC)
+            metadata.update(
+                {
+                    'publishedBy': Version.published_by(now),
+                    'datePublished': now.isoformat(),
+                }
+            )
         return metadata
 
+    @factory.post_generation
+    def assets(self, create: bool, extracted: list[Asset]) -> None:  # noqa: FBT001
+        if not create or not extracted:
+            return
+        self.assets.add(*extracted)
 
-class DraftVersionFactory(BaseVersionFactory):
-    class Meta:
-        model = Version
 
+class DraftVersionFactory(BaseVersionFactory):
     version = 'draft'
 
 
 class PublishedVersionFactory(BaseVersionFactory):
-    class Meta:
-        model = Version
-
-    @classmethod
-    def _create(cls, *args, **kwargs):
-        version: Version = super()._create(*args, **kwargs)
-        version.doi = f'10.80507/dandi.{version.dandiset.identifier}/{version.version}'
-        now = datetime.datetime.now(datetime.UTC)
-        version.metadata = {
-            **version.metadata,
-            'publishedBy': version.published_by(now),
-            'datePublished': now.isoformat(),
-        }
-        version.save()
-        return version
+    doi = factory.LazyAttribute(
+        lambda self: f'10.80507/dandi.{self.dandiset.identifier}/{self.version}'
+    )
+    status = Version.Status.PUBLISHED
```

Comment:
Previously, `PublishedVersionFactory` wasn't setting the `status` to `PUBLISHED`, but I think this makes sense. Note, I had to fix some tests accordingly.